File: Part_I/chapter_01_the_role_of_algorithms_in_computing/insert_sorting.js
/**
 * @function
 * 
 * @description
 * Insert sort algorithm
 *
 * @param {boolean} `isOrderByDesc` – A flag that indicate the sorting order.
 */
function insertSorting(isOrderByDesc) {
    var rawInput = [31, 41, 59, 26, 41, 58];
    console.log(rawInput);
    console.log('\n');

    if (isOrderByDesc) {
        for (var j = 1; j < rawInput.length; j++) {
            var key = rawInput[j];
            // Insert rawInput[j] into the sorted sequence rawInput[0..j - 1].
            var i = j - 1;
            while (i > -1 && rawInput[i] < key) {
                rawInput[i + 1] = rawInput[i];
                i--;
            }
            rawInput[i + 1] = key;
            console.log(rawInput);
        }
    } else {
        for (var j = 1; j < rawInput.length; j++) {
            var key = rawInput[j],
                i = j - 1;

            while (i > -1 && rawInput[i] > key) {
                rawInput[i + 1] = rawInput[i];
                i--;
            }
            rawInput[i + 1] = key;
            console.log(rawInput);
        }
    }
}

var isOrderByDesc = false;
insertSorting(isOrderByDesc);
// insertSorting(true);

File: Part_I/chapter_01_the_role_of_algorithms_in_computing/insert_sorting.py
#!/usr/bin/python
# Filename: insert_sorting.py

def insert_sort(is_desc=False):
    input_list = [31, 41, 59, 26, 41, 58]  # [5, 2, 4, 6, 1, 3]
    list_length = len(input_list)
    j = 1
    print input_list
    print '\n'

    if is_desc:
        while j < list_length:
            key = input_list[j]
            i = j - 1  # Insert inut_list[j] into the sorted sequence input_list[0..j - 1].
            while (i > -1) and (input_list[i] < key):
                input_list[i + 1] = input_list[i]
                i = i - 1
            input_list[i + 1] = key
            j = j + 1
            print input_list
    else:
        while j < list_length:
            key = input_list[j]
            i = j - 1   # Insert inut_list[j] into the sorted sequence input_list[0..j - 1].
            while (i > -1) and (input_list[i] > key):
                input_list[i + 1] = input_list[i]
                i = i - 1
            input_list[i + 1] = key
            j = j + 1
            print input_list

if __name__ == '__main__':
    # insert_sort()
    insert_sort()
